### DISCUSSION TITLE
Still getting `ImportPathMismatchError` when using tox and fixtures `conftest.py`

### DISCUSSION THREAD

**Discussion by astrojuanlu** [pytest-dev/pytest#10109] [Q&A] [answered]
I am observing a weird situation that makes me question my pytest setup. So far my tests were working well, but today I decided to add some fixtures to `conftest.py` that actually import the library code, and started getting these errors when running tox:

```
_________________________________________________________________________________________ ERROR collecting src/poliastro/core/thrust/change_argp.py __________________________________________________________________________________________
.tox/py3-fast/lib/python3.8/site-packages/pytest_doctestplus/plugin.py:246: in collect                                                                                                                                                        
    module = import_path(fspath, root=self.config.rootpath)                                                                                                                                                                                   
.tox/py3-fast/lib/python3.8/site-packages/_pytest/pathlib.py:556: in import_path                                                                                                                                                              
    raise ImportPathMismatchError(module_name, module_file, path)                                                                                                                                                                             
E   _pytest.pathlib.ImportPathMismatchError: ('poliastro.core.thrust.change_argp', '/home/juanlu/Projects/LSF/poliastro/library/.tox/py3-fast/lib/python3.8/site-packages/poliastro/core/thrust/change_argp.py', PosixPath('/home/juanlu/Projects/LSF/poliastro/library/src/poliastro/core/thrust/change_argp.py')) 
```

There are no `__init__.py` files under my `tests` directory, but I acknowledge that maybe it's a weird interaction between tox, pytest, and pytest-doctestplus. The error disappears if I remove my library imports from `conftest.py`.

This is probably similar to what @matthewfeickert observed in https://github.com/scikit-hep/pyhf/pull/1467 and I solved it in the same way: removing `src` from `testpaths`. I'm aware there's also `PY_IGNORE_IMPORTMISMATCH=1`, but it feels kind of hacky.

Basically, my question is: is this a bug? A feature? Am I doing anything wrong by removing `src` from `testpaths` (like suddenly disabling doctests without noticing)? Is `PY_IGNORE_IMPORTMISMATCH=1` the only way?

**Comment by portyu9** [comment, marked as answer]
Hi @astrojuanlu 

The mismatch is real: pytest is ending up with **two different copies of `poliastro` under the same import name**.

Your `conftest.py` imports the installed copy from the tox environment, while collecting `src` for doctests causes pytest/pytest-doctestplus to encounter the source-tree copy. Once one copy has already been imported, pytest detects that the module being collected comes from a different path and raises `ImportPathMismatchError`.

I would not use `PY_IGNORE_IMPORTMISMATCH=1` as the normal solution. It disables the safety check and can hide the fact that you are testing/importing a different copy than intended.

The preferred fix is to make each pytest invocation use **one package location consistently**:

- If you want to test the installed tox package, don't simultaneously collect the source tree as the same importable package.
- If you specifically need source-tree doctests, run those in an invocation/environment where the source tree is the copy being imported.
- For modern pytest projects, `--import-mode=importlib` is also preferable to the older `prepend` import mode because it avoids many `sys.path`/module-name collisions, although a plugin which explicitly imports the installed package can still create the same fundamental two-copy problem.

So removing `src` from `testpaths` appears to "fix" it because it removes the second collection source, but the underlying issue is the simultaneous installed/source-tree imports.

Please mark this answer as accepted if it helped, thank you.
